### DISCUSSION TITLE
Lesson 12: Change Token Collateral And Deploy But Can't Deposit To DSC Engine

### DISCUSSION THREAD

**Discussion by terranweb3** [Cyfrin/foundry-full-course-cu#642] [Q&A]
Hello guys, I am trying to change wBTC contract in Helper Config and I have deployed contract you can see [here](https://sepolia.etherscan.io/address/0xc3ce97451cce9dead8e1bc4c538b5621639ae458).

I receive WBTC from [Aave Faucet](https://staging.aave.com/faucet/). You can see the WBTC contract [here](https://sepolia.etherscan.io/address/0xf864f011c5a97fd8da79baed78ba77b47112935a). WBTC Address: 0xf864f011c5a97fd8da79baed78ba77b47112935a

I changed the wbtc address to the one above and I tried to deposit WBTC into the DSCEngine contract.

Currently my function depositCollateral is still the same as Patrick:
```
function depositCollateral(
        address tokenCollateralAddress,
        uint256 amountCollateral
    )
        public
        moreThanZero(amountCollateral)
        isAllowedToken(tokenCollateralAddress)
        nonReentrant
    {
        s_collateralDeposited[msg.sender][
            tokenCollateralAddress
        ] += amountCollateral;
        emit ColatteralDeposited(
            msg.sender,
            tokenCollateralAddress,
            amountCollateral
        );
        bool success = IERC20(tokenCollateralAddress).transferFrom(
            msg.sender,
            address(this),
            amountCollateral
        );
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
```

When I send a transaction to the DSCEngine contract I get the error: Fail with error 'ERC20: transfer amount exceeds balance'

You can view this tx [here](https://sepolia.etherscan.io/tx/0x79fb97414535bc16d4198fb2c4856406224282327b24072680f4380667e9f87e).

Thanks for all of your answers.

**Comment by theirrationalone** [comment]
You haven't permission to transfer aave token with enough allowance!.... 

Here below, You tried to transfer aave token with some amount of collateral....  

```solidity
bool success = IERC20(tokenCollateralAddress).transferFrom(
            msg.sender,
            address(this),
            amountCollateral
);
```
Look below, this function does both permission as well as make allowance... 

```solidity
function permit(
    address owner,
    address spender,
    uint256 value,
    uint256 deadline,
    uint8 v,
    bytes32 r,
    bytes32 s
  ) external;
```

Aave Protocol doesn't know who's the spender, owner, value(collateral), for how much time, etc.... 

Please try to approve before transfer with all permit function's required args... You will have to modify all the logic of `DscEngine`. 

Please read all the Aave's doc to implement.

**Reply by terranweb3** [reply]
So is there a way to use another token to test on an actual deploy field? Or do I have to manually generate WETH and WBTC tokens to deploy and then transfer to my personal wallet before I can test it?

**Reply by theirrationalone** [reply]
Bro sorry to say, but if we try to implement someone's implementation then we would have to follow their protocol and that's why Blockchain does Persist to be decentralized where majority rules not monarchy. You can follow their protocols for better understanding...

BTW, You still have One Option remains though, After learning `Lesson 12` You can make/implement/build your own specific protocol/idea....

If people liked your idea/protocol then Congrats! You're An Entrepreneur!!!

**Reply by terranweb3** [reply]
Thanks for your advice. But I really just want to try deploying the DSCEngine contract to interact with but it can be quite difficult.

**Reply by theirrationalone** [reply]
You're Welcome! Bro 😇

**Reply by theirrationalone** [reply]
I appreciate 🚀Your curiosity!
